crypto/conf/conf_mod.c

```diff
@@ -384,18 +398,18 @@ static CONF_MODULE *module_find(const char *name)
     if (!RUN_ONCE(&init_module_list_lock, do_init_module_list_lock))
         return NULL;
 
-    if (!CRYPTO_THREAD_read_lock(module_list_lock))
-        return NULL;
+    ossl_rcu_read_lock(module_list_lock);
+    mods = ossl_rcu_derefrence(&supported_modules);
```

Comment:
This is what confused me. I thought you do not need any read locking with RCU, just atomic pointer load. However that would require refcounting on the referenced object which is impossible to do atomically with the dereference.

So in the end RCU is IMO not the solution for this type of objects where writers are very uncommon and in most cases we just read. We would have to use lockless data structures and disallow removing data from them to really improve the performance of readers.

Reply:
> I thought you do not need any read locking with RCU, just atomic pointer load. However that would require refcounting on the referenced object which is impossible to do atomically with the dereference.

There is no read lock as such (ossl_rcu_read_lock() just atomically increments a counter). I don't understand your point here. No refcounting on the object is necessary?

> So in the end RCU is IMO not the solution for this type of objects where writers are very uncommon and in most cases we just read. We would have to use lockless data structures and disallow removing data from them to really improve the performance of readers.

Huh? @nhorman's numbers seem to indicate that it does perform well in this scenario.

Reply:
That is not quite correct.  RCU doesn't require read side locking per-se, but it does require read side announcement of the referencing of data.  That is to say, the read side has to inform the rcu lock that it is referencing data that is under protection of the lock.  The advantage of RCU is that, unlike read write locks, the read side lock will never block, even in the case where a writer is updating the data.  The announcement is for the benefit of the write side, which, after a copy and update of a shared atomic pointer, has to wait until such time as there are no more readers prior to freeing the old shared pointer.  The ossl_rcu_read_lock/unlock calls define that time boundary for the write side

we could rename ossl_rcu_read_[lock|unlock] to something like ossl_rcu_read_announce/release or some such, but traditionally rcu uses the read_lock/unlock terminology in almost all other implementations, so it seems sensible to keep that convention.

Reply:
OK, that makes sense, thanks for the explanation.